feat: jest testing for typescript react

nictownsend opened the pull request:
- jest project support for running server and client tests
- example Home panel with cucumber tests
- jest support files for RTL
- move tsconfig base config into root (not really dev util)
- update incorrect docs regarding module path/aliasing

Contributes to: strimzi/strimzi-ui#77

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

nictownsend committed:
feat: jest testing for typescript react

- jest project support for running server and client tests
- example Home panel with cucumber tests
- jest support files for RTL
- move tsconfig base config into root (not really dev util)
- update incorrect docs regarding module path/aliasing

Contributes to: strimzi/strimzi-ui#77

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

nictownsend commented on .github/actions/bundle/index.js:
```diff
@@ -47,9 +47,10 @@ const createTableHeader = (tableHeadings) => {
-  bundleSizes[bundle.label].current = bundle.parsedSize;
-  const currentSizeBytes = bundleSizes[bundle.label].current;
-  const masterSizeBytes = bundleSizes[bundle.label].master;
+  const bundleSize = bundleSizes[bundle.label] || {master: 0};
```
Cope for new bundles being produced that aren't yet in upstream.

nictownsend commented on build/webpack.client.dev.js:
```diff
@@ -27,26 +26,18 @@ const {
-  withJSModuleLoader,
   withFontModuleLoader,
   withImageModuleLoader,
+  withTypeScriptModuleLoader,
```
Both modules were duplicate code but one set to test for `.js` and other `.ts` - merged the two as the tsconfig is set to compile js too.

nictownsend commented on build/webpack.client.dev.js:
```diff
@@ -63,14 +54,7 @@ const devSpecificConfig = {
-        configFile: './client/tsconfig.json',
-      }),
-    ],
+    plugins: [withTsconfigPathsPlugin({ configFile: 'client/tsconfig.json' })],
```
Nothing using the relative aliases yet. And if they later do, they'll go into tsconfig now.

nictownsend commented on build/webpack.common.js:
```diff
@@ -39,7 +38,7 @@ const withHTMLPlugin = returnPluginWithConfig(HtmlPlugin, {
   title: UI_TITLE, // HTML title
   favicon: `${IMAGES_DIR}/favicon.ico`, // favicon for this page
-  inject: 'head', // any/all built content linked into HTML head element
+  inject: true,
```
Need standard inject otherwise the script runs before the `body` has rendered. @matthew-chirgwin we still need to discuss this.

matthew-chirgwin replied:
Lets chat - more curiosity than anything else!

matthew-chirgwin replied:
Had a chat - as mentioned, was curious this was a requirement for React. Happy with this change, but @nictownsend was going to timebox if there were options to keep scripts in the `head`

nictownsend commented on build/webpack.common.js:
```diff
@@ -59,14 +58,11 @@ const withWebpackBundleAnalyzerPlugin = returnPluginWithConfig(
+      resource.request = resource.request.replace('carbon', 'patternfly');
     }
-  );
+  });
```
Looks like prettier reformatting here, strangely.

nictownsend commented on client/Panels/Home/index.tsx:
```diff
@@ -2,8 +2,4 @@
-module.exports = {
-  presets: babelPresets,
-};
+export * from './Home';
```
No idea why github thinks this is a rename

nictownsend commented on client/tsconfig.json:
```diff
@@ -1,21 +1,12 @@
-      "Hooks": ["Hooks/"],
-      "Pages": ["Pages/"],
-      "Panels": ["Panels/"]
+      "utils/*": ["../utils/*"]
```
All the other aliases are not needed - as typescript will automatically resolve modules using `baseUrl` for a relative lookup if it can't find the module in `node_modules`.

nictownsend commented on test_common/jest.config.js:
```diff
@@ -37,27 +15,17 @@ const config = {
-    },
-  },
+  collectCoverageFrom: [
+    '(client|server)/**/*.{js,ts,jsx,tsx}',
```
By default, jest gets coverage from all files that have been tested (and files it imports). This also includes files without tests (so that coverage drops to reflect missing testing).

nictownsend commented on test_common/jest.config.js:
```diff
@@ -37,27 +15,17 @@ const config = {
+    '!**/server/*',
+    '!**/Home.tsx',
+  ],
+  projects: [
```
Project support to allow jest to run once, but run the two sets of tests for client/server. This is needed because the client and server have different tsconfigs.

JQrdan reviewed (approved):
LGTM

nictownsend commented on utils/tooling/aliasHelper.js:
```diff
@@ -3,58 +3,14 @@
-};
-
-const relativeMocks = Object.keys(relativeAliases).join('|');
 const ignoredBinaries = ['png', 'svg', 'ico', 'scss'].join('|');
```
Significant dead code now that we don't have need for webpack aliases

matthew-chirgwin commented on docs/Build.md:
````diff
@@ -139,6 +85,26 @@ These options will all be provided to the TerserPlugin via it's constructor.
+export * from 'Component.carbon.js';
+```
+
+By setting `process.env.cl=PATTERNFLY` - webpack will replace `carbon` with `patternfly` in the export statement.
````
Should this be `cl` (ie `export cl=<lib>`o r `cl=<lib> npm run build`)?

matthew-chirgwin commented on docs/Build.md:
````diff
@@ -139,6 +85,26 @@ These options will all be provided to the TerserPlugin via it's constructor.
+- And the component exports the carbon implementation:
+
+```ts
+export * from 'Component.carbon.js';
````
js?

matthew-chirgwin reviewed (changes requested):
Looks good, bar a few nits

nictownsend commented on linting/eslint.js.config.js:
```diff
@@ -1,48 +0,0 @@
-/*
```
Merged into root config

nictownsend commented on linting/eslint.ts.config.js:
```diff
@@ -1,22 +0,0 @@
-/*
```
Merged into root config

nictownsend commented on linting/eslint.config.js:
```diff
@@ -2,7 +2,6 @@
  * Copyright Strimzi authors.
```
Note the file path move; @dlabaj PR had updated in original linting location, moved it over to our main one now

nictownsend commented on linting/lint-staged.config.js:
```diff
@@ -5,14 +5,12 @@
 const lintChecksForAllFiles = ['npm run lint:allfiles', 'npm run lint:format']; // note: the licence tool, run as a part of `allfiles`, will update non staged files as well - this cannot be configured currently.
-const lintChecksForJSFiles = ['npm run lint:js'];
-const lintChecksForTSFiles = ['npm run lint:ts'];
+const lintChecksForSrcFiles = ['npm run lint:src'];
```
@dlabaj PR linted TS with JS - so only one config needed now

nictownsend committed:
chore: review comments

- update build docs
- remove babel from package.json

Contributes to: strimzi/strimzi-u#77

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

matthew-chirgwin reviewed (approved):
LGTM

matthew-chirgwin commented on docs/Architecture.md:
```diff
@@ -135,7 +135,7 @@ This UI makes use of:
 - [Sass](https://sass-lang.com/) for styling. This adds a number of helpful features on top of the core css language to commonise and speed up style implementation
 - Webpack and Typescript for building, bundling, treeshaking and transpiling the UI, and enabling day to day development. See [the build documentation](./Build.md#ui-build) for further details on build choices and setup.
-- [`Barrel files`](https://basarat.gitbook.io/typescript/main-1/barrel) and named exports. Barrel files (when combined with [Webpack aliases](./Build.md#webpack-aliases)) allow for many components/modules/functions/constants which are related to be imported and used via one import statement. This relies on exported components/modules/functions/constants to be individually named so they can be imp ...
+- [`Barrel files`](https://basarat.gitbook.io/typescript/main-1/barrel) and named exports. Barrel files (when combined with [path mapping](./Build.md#module-aliases)) allow for many components/modules/functions/constants which are related to be imported and used via one import statement. This relies on exported components/modules/functions/constants to be individually named so they can be importe ...
```
With the change below referencing `cl` as the environment variable to use, can we update https://github.com/strimzi/strimzi-ui/blob/master/docs/Architecture.md#swappable-view-layers to be the same (`VL` -> `cl`)

matthew-chirgwin reviewed (commented):
Just adding one thing I spotted while rebasing some of my changes

nictownsend committed:
chore: further review

- few changes to docs regarding module swappoing

Contributes to: strimzi/strimzi-ui#77

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

matthew-chirgwin commented:
LGTM - thanks!

nictownsend commented:
@matthew-chirgwin for posterity - https://reactjs.org/docs/add-react-to-a-website.html#step-2-add-the-script-tags - docs put the script tag in `<body>` but we could always look at using on load or a document.createElement in future if we want everything in the `<head>` right now?

matthew-chirgwin commented:
> @matthew-chirgwin for posterity - https://reactjs.org/docs/add-react-to-a-website.html#step-2-add-the-script-tags - docs put the script tag in `<body>` but we could always look at using on load or a document.createElement in future if we want everything in the `<head>` right now?

Agreed - I do think it is cleaner for it all to be in the `<head>`, but this should be pretty transparent to the rest of the client code. Thus, we can re visit if we care enough about it later.
